### PR TITLE
NSH over MAVLink

### LINKED ISSUE #2650: Implement NSH over MAVLink
@tridge has added a wrapper packet to the protocol and implemented it for APM here, and it looks super useful as it deconflicts the USB port and makes it easier to access NSH in general.
https://github.com/diydrones/ardupilot/blob/master/libraries/AP_HAL_PX4/NSHShellStream.cpp

### PR DESCRIPTION
**Pull request by bkueng**
This adds the long overdue mavlink shell.

Adds a script `Tools/mavlink_shell.py`, that can be used to connect via UART, TCP & UDP. With this, we can always start mavlink on the USB UART, independent from SD Card.

Fixes #2650

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@DonLakeFlyer @dogmaphobic Both APM and PX4 implement this and it would be a major improvement for developers to have a proper shell in QGC. It's one of these things where a lot of people seem to struggle with and which is tremendously easier when built into the UI. This also gives a shell via Wifi.

**Comment by LorenzMeier** [comment]
@bkueng Fantastic work, this really advances things considerably in terms of the development workflow!

**Comment by mhkabir** [comment]
@bkueng Many, many thanks! Love this.

File: Tools/mavlink_shell.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+                # Erase stdin output (otherwise we have it twice)
+                # FIXME: this assumes a prompt of 5 chars, and does not work in all cases
+                CURSOR_UP_ONE = '\x1b[1A'
+                CURSOR_RIGHT = '\x1b[5C'
+                ERASE_LINE = '\x1b[2K'
+                ERASE_END_LINE = '\x1b[K'
+                sys.stdout.write(CURSOR_UP_ONE + CURSOR_RIGHT + ERASE_END_LINE)
+
+                # TODO: add a command history?
```
Comment (julianoes) [inline Tools/mavlink_shell.py]: I moved the history functionality of pxh into its own class. It's cpp but you might be able to more or less copy it from there if you want (not that it's hard though :) )

**Comment by julianoes** [comment]
Reviewed, looks really clean! Is stack usage safe?

**Comment by LorenzMeier** [comment]
Rebased and applied.

File: Tools/mavlink_shell.py (hunk 1 above)
Comment (bkueng) [inline Tools/mavlink_shell.py]: Thanks for the tip!

**Comment by bkueng** [comment]
@julianoes Yes I checked the stack usage, NSH uses around 800B.

**Pull request closed without merging**